### ISSUE TITLE
Feature tabbing bypass blocks

### ISSUE THREAD

**Issue by prushforth** [Maps4HTML/Web-Map-Custom-Element#280]
> That's a convention sometimes referred to as "skip links": https://webaim.org/techniques/skipnav/.
> Here's an example: https://genericcomponents.netlify.app/generic-skiplink/demo/index.html.
> 
> It works by using a CSS hack to visually hide the links and unhide them as they recieve focus. There is no standard way to hide something that's intended to be focusable (related discussions, FWIW: https://github.com/w3c/csswg-drafts/issues/560, https://github.com/w3c/css-a11y/issues/13).
> 
> It's indeed helpful, it relates to <span title="Success Criterion">SC</span> [2.4.1 Bypass Blocks](https://www.w3.org/TR/WCAG21/#bypass-blocks). I suppose I can see a standard implementation of it (both as a CSS property and as a default facility in maps) but I think that's also based on assumptions of how users currently interact with maps (and what is keyboard accessible, under which circumstances).

_Originally posted by @Malvoz in https://github.com/Maps4HTML/Web-Map-Custom-Element/issues/264#issuecomment-768606392_

Was discussing with Ahmad, we think having a forward/backward bypass block mechanism built into the (popup) UI would be helpful / essential.  When you tab to a feature, then tab into its (popup) UI, the tab sequence could include a fast forward/backward affordance to get to the end of the feature tab sequence quickly.  So the "normal" tab-through-features sequence would be interruptible by drilling into one feature and activating its bypass block. 

![image](https://user-images.githubusercontent.com/2437285/106635416-cca2ae00-654e-11eb-9c6b-d01357c2e834.png)

**Comment by Malvoz** [comment]
I'm not sure I understand the description of the solution. Is the problem: 1) focus a feature 2) press <kbd>Enter</kbd> or <kbd>Space</kbd> to open popup 3) next tab focuses the next feature instead of the popup (or content within the popup)? If I understood correctly, and the implementation relies on current Leaflet functionality, the upstream issue is https://github.com/Leaflet/Leaflet/issues/2199 (tracked in https://github.com/Maps4HTML/Web-Map-Custom-Element/issues/247).

**Comment by prushforth** [comment]
> 1) focus a feature 2) press Enter or Space to open popup 3) next tab focuses the next feature instead of the popup (or content within the popup)? 

That's a separate problem that we agreed I will raise an issue on. 

The issue here is there may be a lot of features to tab through, and having a shortcut forward or backwards to the rest of the DOM content after or before the map might be useful.

**Comment by ahmadayubi** [comment]
Within the custom element you can create a hack to focus the popup followed by the next feature but this hack isn't going to change the DOM structure leaflet generates, where they have the popups in a separate div after all the features in the DOM. It's going to be using JavaScript rather than DOM order, ideally it wouldn't need JavaScript but that is on Leaflet's end, and a big change on their part (changing their internal structure and all).

**Comment by Malvoz** [comment]
A Leaflet user implemented a similar hack, albeit in JQuery: http://melmo.github.io/accessibility/code/dist/leaflet.html (the article mentions other non-related or tangentially related issues, some of which Leaflet has fixed already).

 I think Leaflet rather go that route rather than changing the DOM structure since that may not be backwards compatible with people's code.

**Comment by ahmadayubi** [comment]
https://user-images.githubusercontent.com/55214462/106777462-0341fc80-6613-11eb-8de3-6fe43e3e08d5.mp4

Going to add these changes to PR #270 and link this issue in there also.

**Comment by prushforth** [comment]
To me, this looks close to what I think could be a good UI, but I'd appreciate others comments:

![image](https://user-images.githubusercontent.com/2437285/106799607-748ea900-662d-11eb-8abb-f0ec4d53eedf.png)

I think the "Skip to the end" icons should be something like this (left/right as appropriate obviously):

![image](https://user-images.githubusercontent.com/2437285/106800177-20d08f80-662e-11eb-9fb7-2ee2fc392f69.png)

I think there should be a "1 / n" output indicator, where "n" will always be  1 when tabbing through the feature set, but "n" could be many features when navigating a set of features returned by a query from a server.  I think there should be right-pointing arrow beside or above, but associated to the "1 / n" readout, that when activated goes to either the next feature of the "n" or the next feature in the tab order".

**Comment by Malvoz** [comment]
"skip forwards" and "skip backwards" are very generic descriptions, they're pretty confusing to me. I don't think enough context is provided.

**Comment by prushforth** [comment]
> 
> ![image](https://user-images.githubusercontent.com/2437285/106801544-e5cf5b80-662f-11eb-9c43-1edba49ec973.png)

The double chevron with a vertical bar often signifies to skip past everything in the sequence.  I took "bypass block" to mean effectively this - the user doesn't want to have to be trapped in a long sequence of tabbing, so if they open the popup on any feature, they will be placed (on one of the affordances) in the bypass block, allowing them an early exit from the tab sequence, but only being accessible by opening the / any focused feature.  On the other hand, maybe they want, having looked / listened briefly to the content of the focused feature, to go to the next feature in the tab sequence. For that, they use the single right- or left-pointing arrow head (looks like a 'play' icon except is often used for scrolling through a sequence of things, hence pointing in both directions).  

A challenge here is to leverage the same user interface for query results as for feature tab sequences.

>"skip forwards" and "skip backwards" are very generic descriptions, they're pretty confusing to me. I don't think enough context is provided.

**Comment by Malvoz** [comment]
> The double chevron with a vertical bar often signifies to skip past everything in the sequence. I took "bypass block" to mean effectively this - the user doesn't want to have to be trapped in a long sequence of tabbing

Yes that is correct. Skip links are usually descriptive of _what_ it's taking the user to, so for example the most common implementation of skip links is "Skip to main content" (`<main>`) to bypass the site navigation links. But these labels don't identify _the thing_.

And the fact that they're the first thing encountered in a popup may also be confusing (or even that they're inside the popups, although I don't think there is a better place) because the user has just made an action to explore details for a particular feature.

**Comment by Malvoz** [comment]
I think moving `mapml-focus-buttons` inside `leaflet-popup-content-wrapper` and after `leaflet-popup-content` in the DOM would be the most logical, I think, it also doesn't require the user to shift-tab. So the user would a) open the popup b) get the details for the feature (highly dependent on popup container recieving focus, see the [example](http://melmo.github.io/accessibility/code/dist/leaflet.html#goodmap) I linked to in https://github.com/Maps4HTML/Web-Map-Custom-Element/issues/280#issuecomment-772611623 for screen readers to announce the content) c) tab to the "focus buttons" d) tab to "close button" (if that's re-introduced).

**Comment by ahmadayubi** [comment]
Originally it was inside the content-wrapper but leaflet internally disables interaction with content inside the wrapper, so I had to move it out of there and into the container itself.

**Comment by ahmadayubi** [comment]
Latest version:

https://user-images.githubusercontent.com/55214462/106938267-c0ecee00-66ec-11eb-9e29-a47b662e92e6.mp4
